compiler/rustc_codegen_llvm/src/back/archive.rs

```diff
@@ -174,6 +187,71 @@ impl<'a> ArchiveBuilder<'a> for LlvmArchiveBuilder<'a> {
             self.config.sess.fatal(&format!("failed to build archive: {}", e));
         }
     }
+
+    fn inject_dll_import_lib(
+        &mut self,
+        lib_name: Symbol,
+        dll_imports: &[rustc_codegen_ssa::DllImport],
+        tmpdir: &MaybeTempDir,
+    ) {
+        let output_path = {
+            let mut output_path: PathBuf = tmpdir.as_ref().to_path_buf();
+            output_path.push(format!("{}_imports", lib_name.to_ident_string()));
+            output_path.with_extension("lib")
+        };
+
+        // we've checked for \0 characters in the library name already
+        let dll_name_z = CString::new(format!("{}.dll", lib_name)).unwrap();
+        // All import names are Rust identifiers and therefore cannot contain \0 characters.
+        // FIXME: when support for #[link_name] implemented, ensure that import.name values don't
+        // have any \0 characters
+        let import_name_vector: Vec<CString> = dll_imports
+            .iter()
+            .map(|import| CString::new(import.name.to_string()).unwrap())
+            .collect();
+
+        let output_path_z = rustc_fs_util::path_to_c_string(&output_path);
+
+        tracing::trace!("invoking LLVMRustWriteImportLibrary");
+        tracing::trace!("  dll_name {:#?}", dll_name_z);
+        tracing::trace!("  output_path {}", output_path.display());
+        tracing::trace!(
+            "  import names: {}",
+            dll_imports.iter().map(|import| import.name.to_string()).collect::<Vec<_>>().join(", "),
+        );
+
+        use crate::llvm::LLVMRustCOFFShortExport;
+        let ffi_exports: Vec<LLVMRustCOFFShortExport> = import_name_vector
+            .iter()
+            .map(|name_z| LLVMRustCOFFShortExport::from_name(name_z.as_ptr()))
+            .collect();
+        let result = unsafe {
+            crate::llvm::LLVMRustWriteImportLibrary(
+                dll_name_z.as_ptr(),
+                output_path_z.as_ptr(),
+                ffi_exports.as_ptr(),
+                ffi_exports.len(),
+                llvm_machine_type(&self.config.sess.target.cpu),
```

Comment:
```suggestion
                llvm_machine_type(&self.config.sess.target.arch),
```
Then `x86_64` would map to `IMAGE_FILE_MACHINE_AMD64`, `x86` to `IMAGE_FILE_MACHINE_I386`, `aarch64` to `IMAGE_FILE_MACHINE_ARM64`, and `arm` to `IMAGE_FILE_MACHINE_(ARM,ARMNT,THUMB)`.
(Looks like that's all Windows targets that we are supporting.)

I think it's also better to introduce the COFF machine type enum instead of hardcoding the numbers in code.